### DISCUSSION TITLE
Deactivate or de-init

### DISCUSSION THREAD

**Discussion by tony-sol** [pyenv/pyenv#2743] [Q&A] [answered]
I've just started to use pyenv (installed via homebrew) and have a question about "activating/deactivating"

According to docs, i set up `PYENV_ROOT`, but following xdg_spec, like
```bash
export XDG_DATA_HOME="${XDG_DATA_HOME:-~/.local/share}"
export PYENV_ROOT="${XDG_DATA_HOME}/pyenv"
```
That's ok. Next, i should add 
```bash
eval "$(eval pyenv -i)" # or --path
```
into my .zshrc/.bashrc files and it's a bit tricky - i already have latest python installed (via homebrew either) and wanna keep it as global and be able to switching on/off pyenv'ed python.

E.g.
```bash
~ ✔ $ which python3
/opt/homebrew/bin/python3
~ ✔ $ eval "$(pyenv init -)"
~ ✔ $ which python3
/Users/tonysol/.local/share/pyenv/shims/python3
```
what command should i run next to exec just
```bash
~ ✔ $ PATH="$(bash --norc -ec 'IFS=:; paths=($PATH);
for i in ${!paths[@]}; do
if [[ ${paths[i]} == "''/Users/tonysol/.local/share/pyenv/shims''" ]]; then unset '\''paths[i]'\'';
fi; done;
echo "${paths[*]}"')"
```
part of init command?

UPD: ofk it can be alias, but i'm looking for 'native' solution :)

UPD: looks like is `pyenv init --path`, but without [this](https://github.com/pyenv/pyenv/blob/master/libexec/pyenv-init#L218) call, is what i want as `pyenv de-init` or something

**Comment by native-api** [comment]
See the [installation section in the README](https://github.com/pyenv/pyenv#installation).

To follow the XDG spec, [set `PYENV_ROOT` to `~/.local/opt/pyenv`](https://github.com/pyenv/pyenv/issues/1789#issuecomment-834371980).

**Reply by tony-sol** [reply]
Looks like this is not what my question about

**Comment by native-api** [comment]
```sh
$ which python3
/Users/tonysol/.local/share/pyenv/shims/python3
```

This is fine. Unless you specifically select a Pyenv-installed Python version, shims act as if they weren't on PATH.

Use `pyenv which <executable>` to see what real executable would be run when you run the `<executable>` shim.

[A shim runs whichever Python is currently configured](https://github.com/pyenv/pyenv#understanding-shims). [The default is `system` which means to run whatever is further on PATH](https://github.com/pyenv/pyenv#understanding-python-version-selection).

**Reply by tony-sol** [reply]
And again - this is not what my question about.

Long story short - which pyenv's command will remove pyenv's shims from PATH?

**Reply by tony-sol** [reply]
Maybe i should describe what i want to achieve to solve XY problem :)

I want to **auto-init** pyenv shims - load them into PATH if directory has `.python-version` file and **auto-unload** if directory has no `.python-version` file

**Reply by native-api** [reply]
We did not implement this way -- because directory change cannot be intercepted in all cases. E.g. if you change it inside a program.

Perhaps if you explain what you're trying to achieve by that, we'll be able to suggest some solution.

**Reply by tony-sol** [reply]
I want to achieve same functionality as [direnv](https://github.com/direnv/direnv) provides for `.env` files, but for `.python-version` files.

Usecase:
you have a:
- system python with all it's tooling you want to use across system, like - `pip3` and `poetry`
- projects you are develop, which require other python and it's tooling versions, defined if `.python-version` and `pyproject.toml`

So, if you don't push shims into PATH, you have something like this:
```bash
~ ✔ $ which python3; python3 --version
/usr/bin/python3
Python 3.11.4
~ ✔ $ which poetry
/usr/local/bin/poetry
Poetry (version 1.5.1)
```

Next step - install python and poetry required versions for specific project using pyenv
```bash
~ ✔ $ cd ~/Projects/prj1
~/Projects/prj1 ✔ $ cat .python-version
3.9.16
~/Projects/prj1 ✔ $ pyenv install 3.9.16
~/Projects/prj1 ✔ $ eval "$(pyenv init -)"
~/Projects/prj1 ✔ $ pip3 install "poetry==1.4.2" # pip3 is already from shims
~/Projects/prj1 ✔ $ which python3; python3 --version
/Users/tonysol/.local/share/pyenv/shims/python3
Python 3.9.16
~/Projects/prj1 ✔ $ which poetry; poetry --version
/Users/tonysol/.local/share/pyenv/shims/poetry
Poetry (version 1.4.2)
```

Now, after `poetry install` - all correctly work for prj1.
Next, move to other project:
```bash
~/Projects/prj1 ✔ $ cd ../prj2
~/Projects/prj2 ✔ $ cat .python-version
cat: .python-version: No such file or directory   # !!
~/Projects/prj2 ✔ $ which python3; python3 --version
/Users/tonysol/.local/share/pyenv/shims/python3   # Here, i need a /usr/bin/python3
Python 3.9.16
~/Projects/prj2 ✔ $ which poetry; poetry --version  # And here i need a /usr/local/bin/poetry 
/Users/tonysol/.local/share/pyenv/shims/poetry
Poetry (version 1.4.2)
```

This behaviour can be achieve, using shell hooks, like `direnv` does:
```bash
~ ✘ $ direnv hook zsh
_direnv_hook() {
  trap -- '' SIGINT;
  eval "$("/opt/homebrew/bin/direnv" export zsh)";
  trap - SIGINT;
}
typeset -ag precmd_functions;
if [[ -z "${precmd_functions[(r)_direnv_hook]+1}" ]]; then
  precmd_functions=( _direnv_hook ${precmd_functions[@]} )  # !!
fi
typeset -ag chpwd_functions;
if [[ -z "${chpwd_functions[(r)_direnv_hook]+1}" ]]; then
  chpwd_functions=( _direnv_hook ${chpwd_functions[@]} )    # !!
fi
```
We may use `eval "$(pyenv init --path)"` in similar code to **activate** shims in path when it needed, but we don't have code to **deactivate** shims when not needed ones

**Reply by native-api** [reply, marked as answer]
There's no need to "deinit" in the above-described scenario; you only need to replace `which` with `pyenv which` to see what real executable would be run.

That's nevertheless possible with a plugin.
Create an executable `$PYENV_ROOT/plugins/<plugin-name>/bin/pyenv-sh-deinit`.
It will be called when executing `pyenv deinit` when `pyenv` is installed as a shell function.
It should output commands to be executed in the current shell. Those need to do the reverse of what `pyenv init` does as per https://github.com/pyenv/pyenv#advanced-configuration .
It should also handle the `--complete` argument, printing possible continuations for shell completion (none in this case, most probably).

**Reply by tony-sol** [reply]
Oh, ooooooh, i got it now
Well, thats was a really stupid question this way
Thank you so much! ❤️
